src/archive/src/main.rs

```diff
@@ -0,0 +1,394 @@
+        candid::decode_one::<ArchiveConfig>(&bytes).expect("failed to decode log options")
+    }
+}
+
+impl Default for ArchiveConfig {
+    fn default() -> Self {
+        Self {
+            ii_canister: Principal::from_text("rdmx6-jaaaa-aaaaa-aaadq-cai").unwrap(),
+            max_entries_per_call: 1000,
+            last_upgrade_timestamp: 0,
+        }
+    }
+}
+
+/// Index key for the user index.
+#[derive(Debug)]
+struct UserIndexKey {
+    user_number: UserNumber,
+    timestamp: Timestamp,
+    log_index: u64,
+}
+
+impl Storable for UserIndexKey {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        let mut buf = Vec::with_capacity(USER_INDEX_KEY_LENGTH);
+        buf.extend(&self.user_number.to_le_bytes());
+        buf.extend(&self.timestamp.to_le_bytes());
+        buf.extend(&self.log_index.to_le_bytes());
+        assert_eq!(buf.len(), USER_INDEX_KEY_LENGTH);
+        Cow::Owned(buf)
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        assert_eq!(bytes.len(), USER_INDEX_KEY_LENGTH);
+        UserIndexKey {
+            user_number: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[0..8]).expect("failed to read user_number"),
+            ),
+            timestamp: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[8..16]).expect("failed to read timestamp"),
+            ),
+            log_index: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[16..]).expect("failed to read log_index"),
+            ),
+        }
+    }
+}
+
+#[update]
+fn write_entry(user_number: UserNumber, timestamp: Timestamp, entry: ByteBuf) {
+    with_config(|config| {
+        if config.ii_canister != caller() {
+            trap(&format!(
+                "Only {} is allowed to write entries.",
+                config.ii_canister
+            ))
+        }
+    });
+    let idx = with_log(|log| {
+        log.append(entry.as_ref())
+            .expect("failed to append log entry")
+    });
+
+    with_user_index(|index| {
+        let key = UserIndexKey {
+            user_number,
+            timestamp,
+            log_index: idx as u64,
+        };
+
+        // the only way this expect can trigger is when the key size is wrong.
+        // On other failures (e.g. no more stable memory available) the underlying StableBTreeMap will panic directly.
+        index.insert(key, ()).expect("bug: key size mismatch");
+    })
+}
+
+#[query]
+fn get_logs(index: Option<u64>, limit: Option<u16>) -> Logs {
+    let num_entries = sanitize_limit(limit);
+
+    with_log(|log| {
+        let length = log.len();
+        let start_idx = match index {
+            None => length.saturating_sub(num_entries),
+            Some(idx) => idx as usize,
+        };
+
+        let next_idx = if start_idx + num_entries < length {
+            Some((start_idx + num_entries) as u64)
+        } else {
+            None
+        };
+
+        let mut entries = Vec::with_capacity(log.len().min(num_entries));
```

Comment:
Why build the `Vec` with capacity `min(log_size, num_entries)`?

Reply:
This was there for the case when there are less than `num_entries` in the log total. But since this will only be true for a very short time (and even then it's just an irrelevant optimization), I have removed it again.